src/main/java/edu/umich/tl/analytics/CanvasCourseReportServlet.java

```diff
@@ -46,9 +51,12 @@
 	private static final String ACTION_GET_COURSES = "getCoursesPublished";
 	private static final String ACTION_GET_TERMS = "getEnrollmentTerms";
 	private static final String EMAIL_SUFFIX="@umich.edu";
+	private static final String ACTION_POLLING = "polling";
+	private static final String WORKING = "working";
 	private String canvasToken=CanvasCourseReportFilter.canvasToken;
 	private String canvasURL=CanvasCourseReportFilter.canvasURL;
 	ResourceBundle props = ResourceBundle.getBundle("coursereport");
+	ConcurrentHashMap<Thread,CourseReportTask> courseReportReqThreads=new ConcurrentHashMap<Thread,CourseReportTask>();
```

Comment:
Agreed.  Since you can't guarantee the infrastructure will be the way you expect it to be, it's best to not store the map in memory.  It may be more work, but it should be stored elsewhere.  Maybe the session is a more appropriate place?  Or perhaps it will need to be stored in a DB.